ProjFS.Mac/PrjFSKextTests/KauthHandlerTests.mm

```diff
@@ -243,4 +285,94 @@ - (void)testShouldHandleVnodeOpEvent {
     XCTAssertEqual(kauthResult, KAUTH_RESULT_DEFER);
 }
 
+- (void)testHandleVnodeOpEvent {
+    // Setup
+    kern_return_t initResult = VirtualizationRoots_Init();
+    XCTAssertEqual(initResult, KERN_SUCCESS);
+
+    // Parameters
+    const char* repoPath = "/Users/test/code/Repo";
+    const char* filePath = "/Users/test/code/Repo/file";
+    const char* dirPath = "/Users/test/code/Repo/dir";
+    vfs_context_t _Nonnull context = vfs_context_create(NULL);
+    //PerfTracer perfTracer;
+    PrjFSProviderUserClient dummyClient;
+    pid_t dummyClientPid=100;
+
+    // Create Vnode Tree
+    shared_ptr<mount> testMount = mount::Create();
+    shared_ptr<vnode> repoRootVnode = testMount->CreateVnodeTree(repoPath, VDIR);
+    shared_ptr<vnode> testFileVnode = testMount->CreateVnodeTree(filePath);
+    shared_ptr<vnode> testDirVnode = testMount->CreateVnodeTree(dirPath, VDIR);
+
+    // Register provider for the repository path (Simulate a mount)
+    VirtualizationRootResult result = VirtualizationRoot_RegisterProviderForPath(&dummyClient, dummyClientPid, repoPath);
+    XCTAssertEqual(result.error, 0);
+    vnode_put(s_virtualizationRoots[result.root].rootVNode);
+    
+    // Read a file that has not been hydrated yet
+    expectedMessageType[0] = MessageType_KtoU_HydrateFile;
+    testFileVnode->attrValues.va_flags = FileFlags_IsEmpty | FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_READ_DATA,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+    XCTAssertTrue(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+    MockCalls::Clear();
+    callCount = 0;
+
+    // Ensure a file without an IsEmpty tag is not hydrated
+    testFileVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_READ_DATA,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+    XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+    MockCalls::Clear();
+
+    // Ensure a file is Deleted
+    expectedMessageType[0] = MessageType_KtoU_NotifyFilePreDelete;
+    testFileVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_DELETE,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+    XCTAssertTrue(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+    MockCalls::Clear();
+    callCount = 0;
+
+    // Ensure a directory is Deleted
+    expectedMessageType[0] = MessageType_KtoU_NotifyDirectoryPreDelete;
+    expectedMessageType[1] = MessageType_KtoU_RecursivelyEnumerateDirectory;
```

Comment:
@pmj FYI
If we verify DidCallFunction with args it could come up as too separate functions.

Reply:
Yep, I fully intend to be able to check call sequences; maybe not in the first iteration though.